go/src/lib.rs

```diff
@@ -1,9 +1,180 @@
+/// * The `connection_error_message` pointer in the returned `ConnectionResponse` must live until the returned `ConnectionResponse` pointer is passed to `free_connection_response`.
+/// * Both the `success_callback` and `failure_callback` function pointers need to live while the client is open/active. The caller is responsible for deallocating both callbacks.
+// TODO: Consider making this async
+#[no_mangle]
+pub unsafe extern "C" fn create_client(
+    connection_request_bytes: *const u8,
+    connection_request_len: usize,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes =
+        unsafe { std::slice::from_raw_parts(connection_request_bytes, connection_request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            connection_error_message: CString::into_raw(CString::new(err).unwrap()),
+        },
+        Ok(client) => ConnectionResponse {
+            conn_ptr: Box::into_raw(Box::new(client)) as *const c_void,
+            connection_error_message: std::ptr::null(),
+        },
+    };
+    Box::into_raw(Box::new(response))
+}
+
+/// Closes the given `GlideClient`, deallocating it from the heap.
+///
+/// `client_container_ptr` is a pointer to a valid `GlideClient` returned in the `ConnectionResponse` from `create_client`.
+///
+/// # Panics
+///
+/// This function panics when called with a null `client_container_ptr`.
+///
+/// # Safety
+///
+/// * `close_client` can only be called once per client. Calling it twice is undefined behavior, since the address will be freed twice.
+/// * `close_client` must be called after `free_connection_response` has been called to avoid creating a dangling pointer in the `ConnectionResponse`.
+/// * `client_container_ptr` must be obtained from the `ConnectionResponse` returned from `create_client`.
+/// * `client_container_ptr` must be valid until `close_client` is called.
+// TODO: Ensure safety when command has not completed yet
+#[no_mangle]
+pub unsafe extern "C" fn close_client(client_container_ptr: *const c_void) {
+    assert!(client_container_ptr.is_null());
+    drop(unsafe { Box::from_raw(client_container_ptr as *mut ClientContainer) });
+}
+
+/// Deallocates a `ConnectionResponse`.
+///
+/// This function also frees the contained error. If the contained error is a null pointer, the function returns and only the `ConnectionResponse` is freed.
+///
+/// # Panics
+///
+/// This function panics when called with a null `ConnectionResponse` pointer.
+///
+/// # Safety
+///
+/// * `free_connection_response` can only be called once per `ConnectionResponse`. Calling it twice is undefined behavior, since the address will be freed twice.
+/// * `connection_response_ptr` must be obtained from the `ConnectionResponse` returned from `create_client`.
```

Comment:
nit: cross links in docstrings. Please add them if you find it is reasonable.
```suggestion
/// * `connection_response_ptr` must be obtained from the `ConnectionResponse` returned from [`create_client`](create_client).
```